### PR TITLE
 feat(ingest): add looker meta extractor support in sql parsing

### PR DESCRIPTION
**Pull request by sagar-salvi-apptware**
## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-ingestion/src/datahub/ingestion/source/common/resource.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+from datahub.utilities.str_enum import StrEnum
+
+
+class ResourceType(StrEnum):
+    USER_INFO = "USER_INFO"
+    CHANNEL_INFO = "CHANNEL_INFO"
+    USER_ID_MAPPING = "USER_ID_MAPPING"
+
+
+def generate_user_id_mapping_resource_urn(
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/common/resource.py:18]: Can get rid of most of these if we use `PlatformResourceKey` and `PlatformResource` from [here](https://github.com/datahub-project/datahub/pull/11541/files#)

File: metadata-ingestion/src/datahub/ingestion/source/looker/looker_common.py [hunk 2]
```diff
@@ -1631,4 +1641,32 @@ def get_by_id(self, id_: str) -> Optional[LookerUser]:
...
 
         looker_user = LookerUser.create_looker_user(raw_user)
+        if looker_user.email:
+            self._user_email_cache[id_] = looker_user.email
         return looker_user
+
+    def to_platform_resource(
+        self, platform_instance: Optional[str], env: str
+    ) -> Iterable[MetadataChangeProposalWrapper]:
+        resource_urn = generate_user_id_mapping_resource_urn(
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/looker/looker_common.py:1651]: You can use `PlatformResource.create` helper instead

File: metadata-ingestion/src/datahub/sql_parsing/tool_meta_extractor.py [hunk 3]
```diff
@@ -42,14 +55,45 @@ class ToolMetaExtractor:
...
+
+    @classmethod
+    def create(
+        cls,
+        config: ToolMetaExtractorConfig,
+        graph: Optional[DataHubGraph] = None,
+    ) -> "ToolMetaExtractor":
+        looker_user_mapping: Dict[str, str] = {}
+        if graph and config.looker_platform_details:
+            resource_urn = generate_user_id_mapping_resource_urn(
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/sql_parsing/tool_meta_extractor.py:81]: can use `PlatformResource.from_datahub` helper instead

File: metadata-ingestion/src/datahub/sql_parsing/tool_meta_extractor.py [hunk 4]
```diff
@@ -42,14 +55,45 @@ class ToolMetaExtractor:
...
+        self.looker_user_mapping = looker_user_mapping
+
+    @classmethod
+    def create(
+        cls,
+        config: ToolMetaExtractorConfig,
+        graph: Optional[DataHubGraph] = None,
+    ) -> "ToolMetaExtractor":
+        looker_user_mapping: Dict[str, str] = {}
+        if graph and config.looker_platform_details:
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/sql_parsing/tool_meta_extractor.py:80]: Is is mandatory to configure `looker_platform_details` ?  What happens, its not set ?

File: metadata-ingestion/src/datahub/ingestion/source/looker/looker_common.py (hunk 2 above)
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/looker/looker_common.py:1651]: followed by `to_mcps`

File: metadata-ingestion/src/datahub/sql_parsing/tool_meta_extractor.py [hunk 5]
```diff
@@ -42,14 +55,45 @@ class ToolMetaExtractor:
...
+            )
+            aspect = graph.get_aspect_v2(
+                entity_urn=resource_urn,
+                aspect_type=PlatformResourceInfoClass,
+                aspect="platformResourceInfo",
+            )
+            if aspect and aspect.value:
+                with contextlib.suppress(ValueError, AssertionError):
+                    looker_user_mapping = from_serialized_value(aspect.value)
+                    assert isinstance(looker_user_mapping, dict)
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/sql_parsing/tool_meta_extractor.py:94]: can we add some reporting in in looker source about how many users were added to platform resource ? It would also help to have some indication in tool meta extractor report about number of looker users were found in mapping vs not.

File: metadata-ingestion/tests/integration/looker/golden_test_allow_ingest.json [hunk 6]
```diff
@@ -497,6 +497,43 @@
...
+    "entityType": "platformResource",
+    "entityUrn": "urn:li:platformResource:2c0d2b186264ca5f2532539b69f1559e",
+    "changeType": "UPSERT",
+    "aspectName": "platformResourceInfo",
+    "aspect": {
+        "json": {
+            "resourceType": "USER_ID_MAPPING",
+            "primaryKey": "",
+            "value": {
+                "blob": "{}",
```
Comment (mayurinehate) [inline metadata-ingestion/tests/integration/looker/golden_test_allow_ingest.json:526]: In which case would this be empty in real world ? Ideally, we should always receive mappings here.

**Review by mayurinehate — CHANGES_REQUESTED** [review]

File: metadata-ingestion/tests/integration/looker/golden_test_allow_ingest.json (hunk 6 above)
Comment (sagar-salvi-apptware) [reply metadata-ingestion/tests/integration/looker/golden_test_allow_ingest.json:526]: hi @mayurinehate 
In real-world scenarios, mappings are always provided, and I have verified this. However, in the test cases, we don't have user data, which is why the mappings appear empty.

File: metadata-ingestion/src/datahub/sql_parsing/tool_meta_extractor.py (hunk 4 above)
Comment (sagar-salvi-apptware) [reply metadata-ingestion/src/datahub/sql_parsing/tool_meta_extractor.py:80]: Yes we need it for platform-instance

File: metadata-ingestion/src/datahub/sql_parsing/tool_meta_extractor.py (hunk 3 above)
Comment (sagar-salvi-apptware) [reply metadata-ingestion/src/datahub/sql_parsing/tool_meta_extractor.py]: Changed

File: metadata-ingestion/src/datahub/sql_parsing/tool_meta_extractor.py (hunk 4 above)
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/sql_parsing/tool_meta_extractor.py:80]: I believe we would want the defaults work for no platform instance (no additional user configurations) and work as long as we get single platform resource for looker for platform resource type we need. 

We might need to use `PlatformResource.search_by_key` instead of `PlatformResource.from_datahub` for this to work.

File: metadata-ingestion/src/datahub/ingestion/source/looker/looker_common.py [hunk 7]
```diff
@@ -1646,27 +1666,32 @@ def get_by_id(self, id_: str) -> Optional[LookerUser]:
...
 
     def to_platform_resource(
-        self, platform_instance: Optional[str], env: str
+        self, platform_instance: Optional[str]
     ) -> Iterable[MetadataChangeProposalWrapper]:
-        resource_urn = generate_user_id_mapping_resource_urn(
-            LOOKER, platform_instance, env
-        )
+        try:
+            new_platform_resource = LookerPlatformResource(
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/looker/looker_common.py:1672]: We can use `PlatformResource.create` directly and don't really need LookerPlatformResource class

File: metadata-ingestion/src/datahub/ingestion/source/looker/looker_common.py [hunk 8]
```diff
@@ -1646,27 +1666,32 @@ def get_by_id(self, id_: str) -> Optional[LookerUser]:
...
-        )
+            logger.info(
+                f"Looker Users added to the platform resource: {', '.join([ f'{user_id}:{email}' for user_id, email in self._user_email_cache.items()])}"
+            )
 
-        yield from MetadataChangeProposalWrapper.construct_many(
-            resource_urn, aspects=[dpi, resource_info]
-        )
+            for mcp in platform_resource.to_mcps():
+                yield mcp
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/looker/looker_common.py:1685]:
```suggestion
            yield from platform_resource.to_mcps():
```

File: metadata-ingestion/src/datahub/ingestion/source/looker/looker_common.py [hunk 9]
```diff
@@ -1646,27 +1666,32 @@ def get_by_id(self, id_: str) -> Optional[LookerUser]:
...
+
+        except Exception as exc:
+            self.report.report_failure(
+                message="Failed to generate platform resource for looker",
+                context=", ".join(
+                    [
+                        f"{user_id}:{email}"
+                        for user_id, email in self._user_email_cache.items()
+                    ]
+                ),
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/looker/looker_common.py:1695]: Let's not add this to context please.

File: metadata-ingestion/src/datahub/sql_parsing/tool_meta_extractor.py (hunk 4 above)
Comment (sagar-salvi-apptware) [reply metadata-ingestion/src/datahub/sql_parsing/tool_meta_extractor.py]: Removed

File: metadata-ingestion/src/datahub/sql_parsing/tool_meta_extractor.py [hunk 10]
```diff
@@ -73,24 +69,40 @@ def __init__(self, looker_user_mapping: Optional[Dict[str, str]] = None) -> None
...
+                )
+            ]
+
+            if platform_resources and sum(1 for _ in platform_resources) > 1:
+                report.warning(
+                    "Tool metadata extraction failed due to multiple platform resources found for Looker."
+                )
+                raise ValueError(
+                    "Exactly one looker platform resource should be present"
+                )
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/sql_parsing/tool_meta_extractor.py:102]: Why not only check if len(platform_resources)>1?

This should not be fatal error and rest of tool extractor flow should work as expected. Reporting warning is good enough.

"Tool metadata extraction failed due to multiple platform resources found for Looker."  => "Looker user metadata extraction failed. Found more than one looker user id mappings."

File: metadata-ingestion/src/datahub/sql_parsing/tool_meta_extractor.py [hunk 11]
```diff
@@ -100,6 +112,7 @@ def create(
                     looker_user_mapping = from_serialized_value(
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/sql_parsing/tool_meta_extractor.py:112]: Can you check if SerializedResourceValue.as_raw_json can be used here instead. With platform resource sdk available, module datahub.ingestion.source.common.resource should ideally not be needed.

File: metadata-ingestion/src/datahub/ingestion/source/looker/looker_common.py [hunk 12]
```diff
@@ -1419,6 +1424,7 @@ class LookerDashboardSourceReport(StaleEntityRemovalSourceReport):
     user_resolution_latency: Dict[str, datetime.timedelta] = dataclasses_field(
         default_factory=dict
     )
+    user_cache_file: Optional[str] = None
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/looker/looker_common.py:1427]: Is this used anywhere ?

File: metadata-ingestion/src/datahub/ingestion/source/looker/looker_common.py [hunk 13]
```diff
@@ -1631,4 +1640,31 @@ def get_by_id(self, id_: str) -> Optional[LookerUser]:
...
+            )
+
+            self.report.looker_user_count = len(self._user_email_cache)
+            yield from platform_resource.to_mcps()
+
+        except Exception as exc:
+            self.report.report_failure(
+                message="Failed to generate platform resource for looker",
+                exc=exc,
+            )
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/looker/looker_common.py:1670]:
```suggestion
            self.report.warning(
                message="Failed to generate platform resource for looker id mappings",
                exc=exc,
            )
```

File: metadata-ingestion/src/datahub/sql_parsing/tool_meta_extractor.py [hunk 14]
```diff
@@ -29,7 +37,7 @@ def _get_last_line(query: str) -> str:
 
 
 @dataclass
-class ToolMetaExtractorReport(Report):
+class ToolMetaExtractorReport(SourceReport):
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/sql_parsing/tool_meta_extractor.py:40]: Why is this change needed ?

File: metadata-ingestion/src/datahub/ingestion/source/looker/looker_common.py (hunk 12 above)
Comment (sagar-salvi-apptware) [reply metadata-ingestion/src/datahub/ingestion/source/looker/looker_common.py:1427]: No this is not used i ll remove it

File: metadata-ingestion/src/datahub/sql_parsing/tool_meta_extractor.py (hunk 14 above)
Comment (sagar-salvi-apptware) [reply metadata-ingestion/src/datahub/sql_parsing/tool_meta_extractor.py:40]: This was added as a result of **report.warning()**

File: metadata-ingestion/src/datahub/sql_parsing/tool_meta_extractor.py [hunk 15]
```diff
@@ -42,14 +51,68 @@ class ToolMetaExtractor:
...
+
+                if (
+                    platform_resource
+                    and platform_resource.resource_info
+                    and platform_resource.resource_info.value
+                ):
+                    with contextlib.suppress(ValueError, AssertionError):
+                        value = platform_resource.resource_info.value.as_raw_json()
+                        if value:
+                            looker_user_mapping = value
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/sql_parsing/tool_meta_extractor.py:114]: Let's extract this into method `extract_looker_user_mapping_from_graph and add explicit try..except around to  call to `extract_looker_user_mapping_from_graph` for not to break rest of connector.

File: metadata-ingestion/src/datahub/sql_parsing/tool_meta_extractor.py [hunk 16]
```diff
@@ -78,14 +141,49 @@ def _extract_mode_query(self, entry: QueryLog) -> bool:
...
     def extract_bi_metadata(self, entry: QueryLog) -> bool:
         for tool, meta_extractor in self.known_tool_extractors:
             try:
                 if meta_extractor(entry):
                     self.report.num_queries_meta_extracted[tool] += 1
                     return True
-            except Exception:
-                logger.debug("Tool metadata extraction failed with error : {e}")
+            except Exception as e:
+                logger.debug(f"Tool metadata extraction failed with error : {e}")
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/sql_parsing/tool_meta_extractor.py:186]: Thank you

**Review by mayurinehate — COMMENTED** [review]
Mostly looking good. 
As a final check - can you confirm if platform resource is ingested alright even if `extract_owners` is set to false in looker.

File: metadata-ingestion/src/datahub/ingestion/source/looker/looker_source.py [hunk 17]
```diff
@@ -1152,13 +1152,10 @@ def _get_looker_dashboard(self, dashboard: Dashboard) -> LookerDashboard:
...
-            self.user_registry.get_by_id(
-                user_id,
-            )
-            if self.source_config.extract_owners and user_id is not None
-            else None
-        )
+        if user_id is None:
+            return None
+
+        user = self.user_registry.get_by_id(user_id)
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/looker/looker_source.py:1158]: Is it safe to remove ` self.source_config.extract_owners` check here ? I'm afraid that would ingest owners even if the flag is not set .

How about calling LookerSDK's `all_users` api in LookerUserRegistry init and using that info to populate platform resource as well as first level cache for `get_by_id`.

Comment (sagar-salvi-apptware) [reply metadata-ingestion/src/datahub/ingestion/source/looker/looker_source.py]: Resolved

File: metadata-ingestion/src/datahub/ingestion/source/looker/looker_lib_wrapper.py [hunk 18]
```diff
@@ -154,6 +154,17 @@ def get_user(self, id_: str, user_fields: str) -> Optional[User]:
         # User not found
         return None
 
+    @lru_cache(maxsize=5000)
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/looker/looker_lib_wrapper.py:157]: Please remove caching if not needed

File: metadata-ingestion/src/datahub/ingestion/source/looker/looker_lib_wrapper.py [hunk 19]
```diff
@@ -154,6 +154,17 @@ def get_user(self, id_: str, user_fields: str) -> Optional[User]:
...
+    def all_users(self, user_fields: str) -> Sequence[User]:
+        try:
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/looker/looker_lib_wrapper.py:159]: Update stats to include `all_user_calls` and add `self.client_stats.all_user_calls += 1` here

**Review by mayurinehate — APPROVED** [review]

**Comment by mayurinehate** [comment]
CI failures are unrelated
